aptos-move/block-executor/src/view.rs

```diff
@@ -2595,107 +2662,189 @@ mod test {
+        );
+        assert!(latest_view.get_resource_with_layout_read_set_sequential().is_empty());
+    }
+
+    fn assert_fetch_eq<V: TransactionWrite>(fetched: Option<ValueWithLayout<V>>, expected_maybe_write: Option<V>, expected_maybe_layout: Option<MoveTypeLayout>) {
+        match fetched {
+            Some(ValueWithLayout::Exchanged(write, layout)) => {
+                let expected_write = expected_maybe_write.unwrap();
+                assert_eq!(write.write_op_kind(), expected_write.write_op_kind());
+                assert_eq!(write.bytes(), expected_write.bytes());
+                assert_eq!(write.as_state_value(), expected_write.as_state_value());
+                assert_eq!(layout.as_ref().map(|v| v.as_ref()), expected_maybe_layout.as_ref());
+            },
+            Some(ValueWithLayout::RawFromStorage(_)) => panic!("Unexpected RawFromStorage"),
+            None => {
+                assert_none!(expected_maybe_write);
+                assert_none!(expected_maybe_layout);
+            }
+        }
+    }
+
+    #[test]
+    fn test_sequential_regular_read_operations() {
+        let state_value = create_state_value(&Value::u64(12321), &MoveTypeLayout::U64);
+        let data = HashMap::from([
+            (KeyType::<u32>(1, false), state_value.clone())
+        ]);
+        let h = Holder::new(data, 1000);
+        let latest_view = create_sequential_latest_view(&h, true);
+
+        assert_ok_eq!(
+            latest_view.get_resource_state_value(&KeyType::<u32>(1, false), None),
+            Some(state_value.clone())
+        );
+        assert!(latest_view.get_resource_with_layout_read_set_sequential().is_empty());
+
+        assert_fetch_eq(
+            h.unsync_map.fetch_data(&KeyType::<u32>(1, false)),
+            Some(TransactionWrite::from_state_value(Some(state_value))),
+            None,
+        );
+    }
+
+    #[test_case(Some(true))]
+    #[test_case(Some(false))]
+    #[test_case(None)]
+    fn test_sequential_aggregator_read_operations(check_metadata: Option<bool>) {
+        let layout = create_aggregator_layout();
+        let value = create_aggregator_value(25, 30);
+        let state_value = create_state_value(&value, &layout);
+        let data = HashMap::from([
+            (KeyType::<u32>(1, false), state_value.clone())
+        ]);
+        let h = Holder::new(data, 1000);
+        let latest_view = create_sequential_latest_view(&h, true);
+
+        let patched_value = create_aggregator_value(1000, 30);
+        let patched_state_value = create_state_value(&patched_value, &layout);
+
+        match check_metadata {
+            Some(true) => {
+                latest_view
+                    .get_resource_state_value_metadata(&KeyType::<u32>(1, false))
+                    .unwrap();
+            },
+            Some(false) => {
+                assert_ok_eq!(
+                    latest_view.resource_exists(&KeyType::<u32>(1, false)),
```

Comment:
this is the test - before - when resource_exists gets called, we store None as layout and don't do exchange, and then get_resource_state_value below returns incorrect values.